tests/unit/reco_utils/evaluation/test_spark_diversity_evaluator.py

```diff
@@ -0,0 +1,137 @@
+        ),
+        "user_novelty": pd.DataFrame(
+        dict(UserId=[1, 2, 3], user_novelty=[0.0, 0.0, 0.5])
+        ),
+        "novelty": pd.DataFrame(
+        dict(novelty=[0.16667])
+        ),
+        "diversity": pd.DataFrame(
+        dict(diversity=[0.43096])
+        ),
+        "user_diversity": pd.DataFrame(
+        dict(UserId=[1, 2, 3], user_diversity=[0.29289, 1.0, 0.0])
+        ),
+        "user_item_serendipity": pd.DataFrame(
+        dict(UserId=[1, 1, 2, 2, 3, 3], ItemId= [3, 5, 2, 5, 1, 2], user_item_serendipity=[0.72783, 0.80755, 0.71132, 0.35777, 0.80755, 0.80755])
+        ),
+        "user_serendipity": pd.DataFrame(
+        dict(UserId=[1, 2, 3], user_serendipity=[0.76770, 0.53455, 0.80755])
+        ),
+        "serendipity": pd.DataFrame(
+        dict(serendipity=[0.70326])
+        ),
+    }
+
+
+
+@pytest.fixture(scope="module")
+def data(spark):
+    train_df = spark.createDataFrame([
+      Row(UserId=1, ItemId=1),
+      Row(UserId=1, ItemId=2),
+      Row(UserId=1, ItemId=4),
+      Row(UserId=2, ItemId=3),
+      Row(UserId=2, ItemId=4),
+      Row(UserId=3, ItemId=3),
+      Row(UserId=3, ItemId=4),
+      Row(UserId=3, ItemId=5),  
+    ])
+    reco_df = spark.createDataFrame([
+      Row(UserId=1, ItemId=3, Rating=1),
+      Row(UserId=1, ItemId=5, Rating=1),
+      Row(UserId=2, ItemId=2, Rating=1),
+      Row(UserId=2, ItemId=5, Rating=1),
+      Row(UserId=3, ItemId=1, Rating=1),
+      Row(UserId=3, ItemId=2, Rating=1),  
+    ])
+    return train_df, reco_df
+    
+@pytest.mark.spark    
+@pytest.fixture()
+def evaluator(data):
+    train_df, reco_df = data
+    div = DiversityEvaluator(train_df=train_df, reco_df=reco_df,
+                         user_col='UserId', item_col='ItemId')
+    print("init evaluator")
+    return div
+
+
+@pytest.mark.spark
+def test_init_spark(spark):
+    assert spark is not None
+
+@pytest.mark.spark
+def test_catalog_coverage(evaluator, target_metrics):
+      
+    c_coverage = evaluator.catalog_coverage()
+    assert c_coverage == target_metrics["c_coverage"]
+    
+@pytest.mark.spark
+def test_distributional_coverage(evaluator, target_metrics):
+      
+    d_coverage = evaluator.distributional_coverage()
+    assert d_coverage == target_metrics["d_coverage"]
+
+@pytest.mark.spark
+def test_item_novelty(evaluator, target_metrics):
+    actual = evaluator.item_novelty().toPandas()
+    assert_frame_equal(target_metrics["item_novelty"], actual, check_exact=False, check_less_precise=4)
+
+@pytest.mark.spark    
+def test_user_novelty(evaluator, target_metrics):
+    actual = evaluator.user_novelty().toPandas()
+    assert_frame_equal(target_metrics["user_novelty"], actual, check_exact=False, check_less_precise=4)
+
+@pytest.mark.spark    
+def test_novelty(evaluator, target_metrics):
+    actual = evaluator.novelty().toPandas()
+    assert_frame_equal(target_metrics["novelty"], actual, check_exact=False, check_less_precise=4)
```

Comment:
We also need to check the limits, this is very important to check that the formulas are correct. This would be check perfect novelty and non-novelty, perfect diversity and non-diversity, etc. 

See example: https://github.com/microsoft/recommenders/blob/main/tests/unit/reco_utils/evaluation/test_spark_evaluation.py#L104

Reply:
"check the limits" -- what is the precision you would recommend. I am comparing the number to the 4 positions after the decimal point.

Reply:
4 position is great, we even have 2 position in some cases